### PR TITLE
Boundary integrals only on indicated boundaries [form_marker_dev]

### PR DESCRIPTION
**Pull request by IdoAkkerman**
This adds a boundary marker to the linearform and bilinearform. This allows the integrator to be only active on indicated boundaries. The interfaces and logic are analogous to the face integrals.

### CODE REVIEW

**Comment by tzanio** [comment]
@mlstowell, how does this relate to your changes in the [entity-sets-dev](https://github.com/mfem/mfem/tree/entity-sets-dev) branch?

File: fem/bilinearform.cpp [hunk 1]
```diff
@@ -357,8 +365,32 @@ void BilinearForm::Assemble (int skip_zeros)
...
+      for (int k = 0; k < bbfi.Size(); k++)
+      {
+         if (bbfi_marker[k] == NULL)
+         {
+            bdr_attr_marker = 1;
+            break;
+         }
+         Array<int> &bdr_marker = *bbfi_marker[k];
+         MFEM_ASSERT(bdr_marker.Size() == bdr_attr_marker.Size(),
+                     "invalid boundary marker for boundary face integrator #"
```
Comment (mlstowell) [inline fem/bilinearform.cpp]: Minor edit: I would change this comment from "... boundary face integrator" to "... boundary integrator"

File: fem/linearform.cpp [hunk 2]
```diff
@@ -69,12 +78,38 @@ void LinearForm::Assemble()
...
+      for (int k = 0; k < blfi.Size(); k++)
+      {
+         if (blfi_marker[k] == NULL)
+         {
+            bdr_attr_marker = 1;
+            break;
+         }
+         Array<int> &bdr_marker = *blfi_marker[k];
+         MFEM_ASSERT(bdr_marker.Size() == bdr_attr_marker.Size(),
+                     "invalid boundary marker for boundary face integrator #"
```
Comment (mlstowell) [inline fem/linearform.cpp]: Minor edit: I would change this comment from "... boundary face integrator" to "... boundary integrator"

File: fem/bilinearform.hpp [hunk 3]
```diff
@@ -212,6 +213,11 @@ class BilinearForm : public Matrix
    /// Adds new Boundary Integrator.
    void AddBoundaryIntegrator(BilinearFormIntegrator *bfi);
 
+   /** @brief Adds new Boundary Integrator, restricted to specific boundary
+       attributes. */
+   void AddBoundaryIntegrator(BilinearFormIntegrator *bfii,
```
Comment (mlstowell) [inline fem/bilinearform.hpp]: Minor edit: I would change the argument name from "bfii" to "bfi".

**Comment by mlstowell** [comment]
@tzanio , these changes are similar in spirit to those in the entity-sets-dev branch but I believe both branches would be useful additions to MFEM.  This branch focuses on restricting boundary integrals by attribute which will be useful for building r.h.s. vectors and implementing absorbing boundary conditions.  The entity-sets-dev branch hopes to provide a mechanism for applying Dirichlet boundary conditions to small sets of nodes, edges, faces, or cells and also for post-processing of field data.

**Comment by mlstowell** [comment]
@IdoAkkerman , I apologize for not addressing this pull request sooner.  I think it's a very valuable addition to MFEM!

**Comment by tzanio** [comment]
@mlstowell, did you mean to "approve" this?

File: fem/linearform.cpp [hunk 4]
```diff
@@ -137,7 +137,7 @@ void LinearForm::Assemble()
          }
          Array<int> &bdr_marker = *flfi_marker[k];
          MFEM_ASSERT(bdr_marker.Size() == bdr_attr_marker.Size(),
-                     "invalid boundary marker for boundary face integrator #"
+                     "invalid boundary marker for boundary integrator #"
```
Comment (v-dobrev) [inline fem/linearform.cpp]: The text of the error message was correct: "boundary integrator" is different from "boundary face integrator" and this loop is processing the latter.

**Pull request merged**